Some work on the memory handling

TomTheBear opened the pull request:
- Reduce memory footprint: Use a temporary token during parsing and instead of reallocating a new one for every parsing step.
- Remove frequent reallocs: Initialize stack with maximal size
- Remove all memory leaks: All intermediate tokens are stored and freed in the end
- Add support for scientific notation with negative exponent: Tokens like 1E-09 are now recognized properly
- Minor change: Use higher precision

btmills commented on calculator.c:
```diff
@@ -470,65 +477,76 @@ int tokenize(char *str, char *(**tokensRef))
 				tokens = (char**)malloc(numTokens * sizeof(char*));
 			else*/
-				tokens = (char**)realloc(tokens, numTokens * sizeof(char*));
+			newToken = malloc((strlen(tmpToken)+1) * sizeof(char));
```
Since `newToken` isn't needed until all the way down here, let's move its declaration (line 397) here.

btmills commented on calculator.c:
```diff
@@ -470,65 +477,76 @@ int tokenize(char *str, char *(**tokensRef))
 						 	|| ((*ptr == 'E' || *ptr == 'e') // Or the next character is an exponent
-						 		&& hasExponent == false)); // But we have not added an exponent yet
+						 		&& hasExponent == false) // But we have not added an exponent yet
+						 	|| ((*ptr == '+' || *ptr == '-') && hasExponent == true)); // Exponent with sign
```
Very cool, nice addition :bowtie:

btmills commented on calculator.c:
```diff
@@ -984,6 +1018,14 @@ int main(int argc, char *argv[])
 			free(tokens);
 			tokens = NULL;
 			numTokens = 0;
+			for (i = 0; i < nrCalcTokens; i++)
```
How hard would it be to keep track of this without using global state?

TomTheBear replied:
I tried it but it didn't work out. Moreover, I tried to do as less changes as possible to your code because the version mismatch in LIKWID (BSD vs. GPLv3) requires that I mark all code fragments I included in your code.
But it should be possible to free all tokens that are evaluated in `evalStackPush` and push only the result in a new allocated space on the stack. If I have time, I try that.

btmills commented on calculator.c:
```diff
@@ -366,27 +369,33 @@ Symbol tokenType(token tk)
 			break;
 
 		token newToken = NULL;
+		memset(tmpToken, '\0', MAXTOKENLENGTH+1);
```
How much does zeroing memory impact performance? I'm thinking that since the tokenization process sets all of the bytes through and including the null terminator, it might not need to zero the entire token. Given the `if` on [line 527](https://github.com/btmills/calculator/pull/6/files#diff-0e65cd880e53ce3dd36e6bbae63d6ca5R527), that might still require explicitly zeroing the first byte here.

TomTheBear replied:
I don't think that it harms performance to an extend that the calculator seems slow. I do the memset only to have a consistent state at each tokenization step. It is probably enough to set only the first entry to a null byte for the `if` you mentioned.

btmills commented:
These are great changes, thank you! :sparkles: I had a couple questions, but this looks like it will be a big improvement.

TomTheBear commented:
So, I did some changes to the code.
- There is now no global state anymore. The intermediate calculation results are stored and freed in the `postfix` function.
- The functions can be variadic now. I added `min`, `max`, `sum`, `mean` and `median`. Also the precedence is handled correctly.
- The precision is configurable to `x` decimal places or `auto` (removes trailing zeros). I added it to your settings system, so `set precision [X|auto]` to set it and `get precision` to get the current setting. The default precision is set to 5 decimal places.
- The maximal token length can be set on the command line, default is 512 characters, so long enough for most cases

btmills commented:
This is way cool, thanks @TomTheBear! The calculator is becoming pretty powerful. The new functions are so useful, and I like the configurable precision.

Running through this, I found a few cases that aren't handled well:

- Typing a single number, e.g. `1`, will crash with "pointer being freed was not allocated".
- If I call `min()`, `max()`, `sum()`, or `median()` without any arguments all segfault.
- `mean()` with any number of arguments prints "Error evaluating expression".
- `mean(1)` with only the one argument crashes with "pointer being freed was not allocated".
- `min` and `max` don't seem to handle negative numbers in some cases - `min(3, -1)` gives 2, though `min(-1, 3)` correctly gives -1, and `max(3, -1)` gives 2, whereas `max(-1, 3)` correctly gives 3.
- From a freshly launched instance, if I run `median(1, 2, 3, 4, 5, 6, 7, 8, 9)` twice, it crashes with "incorrect checksum for freed object - object was probably modified after being freed".

Any idea what could be causing those?

TomTheBear commented:
Hi,

I checked the cases you mentioned and found the problematic code parts:

- Typing a single number, e.g. `1`, will crash with "pointer being freed was not allocated".

This reason for this is, that the input token is the output token, a situation that never happens if you have any kind of calculation. I added a small loop that checks if any input token == output token and skip the second `free()` for this token.

- If I call `min()`, `max()`, `sum()`, or `median()` without any arguments all segfault.

This is true, I forgot to check that at least one token must be on the stack when calculating a function. I added a check at the beginning at `doFunc()` and a new error case 'inputMissing'.

- `mean()` with any number of arguments prints "Error evaluating expression".

This is mainly a documentation error. I called the function`avg` and not `mean`, so if you execute `mean` there is no function behind that. This is a failure in my above text. `mean` is now the same as `avg`.

- `mean(1)` with only the one argument crashes with "pointer being freed was not allocated".

Same reason as the first point. Input token `1` is the same as the output token `1` because if there is only one argument to `mean`, `min`, `max` and `median` it just returns this token. No real function evaluation must be done.

- `min` and `max` don't seem to handle negative numbers in some cases - `min(3, -1)` gives 2, though `min(-1, 3)` correctly gives -1, and `max(3, -1)` gives 2, whereas `max(-1, 3)` correctly gives 3.

I checked that and if you have `(3,-1)` it parses the string as `3`, `,`, `-`, `1` and evaluates `3 - 1` before evaluating the function. The check for a negative number was too restrictive, it didn't recognize negative numbers after an argument separator.

- From a freshly launched instance, if I run `median(1, 2, 3, 4, 5, 6, 7, 8, 9)` twice, it crashes with "incorrect checksum for freed object - object was probably modified after being freed".

I cannot reproduce this. 
```
$ ./calc 
median(1, 2, 3, 4, 5, 6, 7, 8, 9)
	= 5.00000
median(1, 2, 3, 4, 5, 6, 7, 8, 9)
	= 5.00000
median(1, 2, 3, 4, 5, 6, 7, 8, 9)
	= 5.00000
median(1, 2, 3, 4, 5, 6, 7, 8, 9)
	= 5.00000
quit
$
```

- In `doFunc()` and `isFunction()`: `strcmp()` is replaced by `strncmp()` for security reasons
- Added variadic function `var` for calculating the population variance

btmills commented:
Nice, that fixed everything! I can no longer reproduce the incorrect checksum error, so I'm assuming that was fixed.

- I'm getting a compiler warning now with a suggested fix:

    ```sh
    $ make
    gcc -c -Wall stack.c -lm
    clang: warning: -lm: 'linker' input unused
    gcc -Wall -o calc calculator.c stack.c -lm
    calculator.c:738:3: warning: variable 'ret' is used uninitialized whenever switch default is taken
          [-Wsometimes-uninitialized]
                    default:
                    ^~~~~~~
    calculator.c:741:9: note: uninitialized use occurs here
            return ret;
                   ^~~
    calculator.c:726:10: note: initialize the variable 'ret' to silence this warning
            bool ret;
                    ^
                     = '\0'
    1 warning generated.
    ```

- Combining functions in other expressions, I'm getting "Error evaluating expression". `1 + max(2, 3)` or `min(1, 2) + max(3, 4)` do this.
- `max(1, min(3, 4))` is giving me `1.00000`

TomTheBear commented:
Interesting, my GCC doesn't throw any warning. I initialize the variable now, so I hope there is no warning anymore.

The reason for the other two errors is that when values are pushed on the output stack for calculation, there is no separation between values that are arguments to a function and arguments to a standard operation. So for for e.g. `1+max(2,3)`, the values `1`, `2`, and `3` are pushed on the output stack and with the right parentheses the `max` function is evaluated on the complete output stack with the result `1`. Afterwards it checks whether there are still operators on the operator stack and finds `+` but there is only one argument left on the output stack (the result of `max`) and prints the error. Therefore, I added a separator that is pushed on the output stack when a function token is found. When parsing the right parentheses the function is evaluated until it finds the separator and finally removes the separator from the output stack before pushing the result to the output stack. All the examples are fixed with that. I also did some more tests and all seem fine.

btmills commented:
Sorry for leaving this for so long, @TomTheBear. I just ran through all the scenarios we've identified up above. Everything is working except `min()`, `max()`, `median()`, `mean()`, and `sum()`, which segfault when called without arguments.

TomTheBear commented:
Hi, no problem, I had enough to do.

I fixed the problem with the variadic functions. The function-arguments-begin-here-sign is pushed on the stack at `(`, so the stack is not empty and the emptiness check fails. I fixed that.

Moreover, I found another segfault when calling `min(1` first (raises Mismatched parentheses) followed by the proper function `min(1)`. This was also fixed.

btmills commented:
Just did another run-through, and everything looks good! Is there anything else you wanted to do in this PR, or is this ready to merge?

TomTheBear commented:
Hi, that's all I want to include for now. Thanks for your patience with all the errors.

btmills commented:
Awesome, thanks for your hard work and patience!
